api: initial implementation of headless API (Bug 1941363)

cgsheeh opened the pull request:
Build out the basic functionality of the headless API for Lando.
Using django-ninja we define two API endpoints, to POST automation
jobs and GET job statuses after submission. The API endpoints take
a set of `actions` defined in the request body which are stored in
the database for processing by a worker. Authentication is handled by
an API key associated with a user profile. A single action, `add-commit`
is implemented which can be used to test adding patches to the repo
as commits.

zzzeid reviewed (changes requested):
Couple of very early comments:
- the new api functionality would be better located in the `lando.api` app not the `lando.main` app. It could also go into a more specific `headless_api` app if we want to separate it from other functionality that was ported or will be ported from the old API.
- I noticed some boilerplate functionality around generating / saving / getting API tokens, as well as some model fields to encrypt / decrypt / store those. Seems like potentially something that should be offloaded to the auth system (and the API framework), and not something that we should manually implement as libraries already exist that do this.
- This PR should be broken up into multiple PRs (e.g., adding API functionality, adding "action" endpoints and worker functionality).

Again very early feedback based on a quick first pass.

cgsheeh commented:
Thanks for looking it over, the PR is still a draft but I wanted to get something up to show progress.


>     * the new api functionality would be better located in the `lando.api` app not the `lando.main` app. It could also go into a more specific `headless_api` app if we want to separate it from other functionality that was ported or will be ported from the old API.

I had originally put the new code in `lando.api` but I ran into some issues with creating new models and referencing the existing models from them. In particular Django didn't like that I was trying to set a foreign key reference to the `Repo` model across apps. In the interest of staying focused on the headless API I moved everything into `lando.main`, but I will look into this further.

>     * I noticed some boilerplate functionality around generating / saving / getting API tokens, as well as some model fields to encrypt / decrypt / store those. Seems like potentially something that should be offloaded to the auth system (and the API framework), and not something that we should manually implement as libraries already exist that do this.

I just re-used the prior art around Phabricator API token storage/retrieval, but I haven't implemented anything for the API token generation. I was expecting to add the token management to the "settings page" where the Phab API tokens are managed.

>     * This PR should be broken up into multiple PRs (e.g., adding API functionality, adding "action" endpoints and worker functionality).

I will try and split out some of the changes around test fixtures, and perhaps the token generation/storage/etc could go in a separate PR. Moving the API definition out of the PR where the worker/job/etc is defined seems like it will make things harder to review since the API will do nothing but write to a DB, but I'll look into it.

shtrom commented on src/lando/main/api.py:
```diff
@@ -0,0 +1,169 @@
+    created_at: datetime.datetime
+
+
+@api.post("/repo/{repo_name}/{branch}", response={202: JobStatus, codes_4xx: ApiError})
```
I don't think we currently have a `branch` separate from the `repo` model. These endpoints should probably follow that, or maybe accepts a naked `/{repo_name}` which would use the default branch.

shtrom commented on src/lando/main/api.py:
```diff
@@ -0,0 +1,169 @@
+        repo = Repo.objects.get(name=repo_name)
+    except Repo.DoesNotExist:
+        error = f"Repo {repo_name} does not exist."
+        logger.info(error)
```
We should enhance the error with something like the username of the requester. Maybe in the `extra` argument https://docs.python.org/3/library/logging.html#logging.Logger.debug?

shtrom reviewed (changes requested):
Looking good. Lots of half-baked musing in my comments. Feel free to consider or discard as you see fit (:

zzzeid commented on src/lando/api/legacy/workers/automation_worker.py:
```diff
@@ -0,0 +1,295 @@
+            return
+
+        with job_processing(job):
+            job.status = LandingJobStatus.IN_PROGRESS
```
Is this a copy/paste error?

cgsheeh replied:
What do you mean by "this"?

zzzeid replied:
this == `LandingJobStatus`, since we are dealing with `AutomationJob` objects in this worker, I expected this to say something like `AutomationJobStatus`.

zzzeid commented on src/lando/main/api.py:
```diff
@@ -0,0 +1,169 @@
+class HeadlessAPIAuthentication(HttpBearer):
+    """Authentication class to verify API token."""
+
+    def authenticate(self, request, token: str) -> str:
```
I think it would be more ideal if we could do proper authentication with only a token (i.e., fetch a user based on a valid token, which is typically what tokens are able to do.)

This could be done by means of adding a tokens table, and you could also possibly leverage Django's `check_password` and `make_password` [functions](https://docs.djangoproject.com/en/5.1/topics/auth/passwords/#module-django.contrib.auth.hashers) for hashing.

zzzeid reviewed (changes requested):
A few more comments related to our discussion on tokens.

shtrom commented on src/lando/main/api.py:
Should this go to src/lando/api/automation.py instead?

I thought it was legacy code, but then there is a legacy subdir for this, so perhaps the rest of this is the right place for API code.

@zzzeid what's your view on this?

shtrom commented on src/lando/api/legacy/workers/automation_worker.py:
```diff
@@ -0,0 +1,295 @@
+        self, job: AutomationJob, repo: Repo, scm: AbstractSCM, action: AddCommitAction
+    ) -> bool:
+        """Run the `add-commit` action."""
+        patch_helper = HgPatchHelper(StringIO(action.content))
+
+        date = patch_helper.get_header("Date")
+        user = patch_helper.get_header("User")
+
+        try:
```
Maybe we could add the `run()` method as suggested in https://github.com/mozilla-conduit/lando/pull/194/files#r1923160292, and then just have the landing_worker also create an `AddCommitAction` and run it, so we have a single codepath for both.

shtrom commented on src/lando/main/api.py:
```diff
@@ -0,0 +1,169 @@
+class HeadlessAPIAuthentication(HttpBearer):
+    """Authentication class to verify API token."""
+
+    def authenticate(self, request, token: str) -> str:
```
I think in practice, a BearerToken is self-standing. It's a plain-text secret (modulo TLS), that can be looked up in a DB column and rotated at will (and may expire), and uniquely identifies a user row without the need for a username.

Here, I think we want a longer-lived token as an API key. Nonetheless, it's probably easier to stick to a Token rather than a User+Password (particularly as we rely on OAuth for user auth in the UI, which would make things murkier).

cgsheeh commented on src/lando/main/models/automation_job.py:
```diff
@@ -0,0 +1,124 @@
+    # Email of the user who created the automation job.
+    requester_email = models.CharField(blank=True, default="", max_length=255)
+
+    # Identifier for the most descendent commit created by this landing.
```
No, "descendant" here is part of the parent/child relationship terminology. A child is a "descendant" of it's parent, so this means the commit which is the most recent in the topography.

cgsheeh commented on src/lando/main/models/automation_job.py:
```diff
@@ -0,0 +1,124 @@
+            missing_params = required_params - kwargs.keys()
+            raise ValueError(f"Missing {missing_params} params")
+
+        self.status = actions[action]["status"]
```
We're not usually supposed to transition from "Landed" or "Failed" to any other status as those are considered final. But there's nothing in the code restricting us from doing so.

cgsheeh commented on src/lando/api/legacy/workers/automation_worker.py:
```diff
@@ -0,0 +1,295 @@
+        repo = job.target_repo
+        scm = repo.scm
+
+        # TODO should we check treestatus?
```
I think there's already an implicit expectation that any in-progress pushes will complete when the `CLOSED` status is set. This comment is more about if the tree should always be considered open and pushes should always be considered on a `CLOSED TREE`, since the set of people who are able to push directly are minimal.

cgsheeh commented on src/lando/api/legacy/workers/automation_worker.py:
```diff
@@ -0,0 +1,295 @@
+            self.throttle(self.sleep_seconds)
+            self.refresh_enabled_repos()
+
+        with transaction.atomic():
+            job = AutomationJob.next_job(repositories=self.enabled_repos).first()
```
This piece is copy/pasted from the landing worker implementation, and we have the same single line there. In the unused "mixin" we have more lines in the atomic context manager... So I'm not entirely sure. :sweat_smile:

cgsheeh commented on src/lando/main/models/automation_job.py:
```diff
@@ -0,0 +1,124 @@
+        AutomationJob, on_delete=models.CASCADE, related_name="actions"
+    )
+
+    action_type = models.CharField()
```
We could use a `ChoiceField`, but the tradeoff is that we would have to do a DB migration when we add a new action type. Perhaps not a big deal, I left it as-is for the moment. Is there any advantage to the `ChoiceField` aside from being more explicit? :)

cgsheeh commented on src/lando/api/legacy/workers/automation_worker.py:
```diff
@@ -0,0 +1,295 @@
+        # Trigger update of repo in Phabricator so patches are closed quicker.
+        # Especially useful on low-traffic repositories.
+        if repo.phab_identifier:
+            self.phab_trigger_repo_update(repo.phab_identifier)
```
The error handling is handled within the callee.

cgsheeh commented on src/lando/main/api.py:
```diff
@@ -0,0 +1,169 @@
+        # some APIs may have authentication without user management. Our
+        # API tokens always correspond to a specific user, so set that on
+        # the request here.
+        request.user = user
```
I have yet to look into this but it is on my radar.

cgsheeh replied:
It looks like `django.contrib.auth.login` is used to log the user in to Django's session framework. Since we are using API tokens here, each request will re-authenticate the user statelessly, so I don't think logging into the session framework is necessary. :+1:

shtrom commented on src/lando/api/legacy/workers/automation_worker.py:
```diff
@@ -0,0 +1,295 @@
+        with scm.for_push(job.requester_email):
+            repo_pull_info = f"tree: {repo.tree}, pull path: {repo.pull_path}"
+            try:
+                # TODO should we always update to the latest pull_path for a repo?
+                # or perhaps we need to specify some commit SHA?
+                scm.update_repo(repo.pull_path)
```
Correction: now it creates a new branch in git. This may be a problem we want to come back to?

shtrom commented on src/lando/headless_api/tests/test_automation_api.py:
```diff
@@ -0,0 +1,427 @@
+    # TODO test a few more things? formatting?
+
+
+PATCH_NORMAL_1 = r"""
+# HG changeset patch
+# User Test User <test@example.com>
+# Date 0 0
+#      Thu Jan 01 00:00:00 1970 +0000
+# Diff Start Line 7
+add another file.
+diff --git a/test.txt b/test.txt
+--- a/test.txt
++++ b/test.txt
+@@ -1,1 +1,2 @@
+ TEST
++adding another line
+""".strip()
```
We can use the `normal_patch` fixture https://github.com/mozilla-conduit/lando/blob/ff0dbd3c06af3830975e3f3e70f6949b9c964938/src/lando/api/tests/conftest.py#L131-L143to get this one out, rather than duplicating it.

shtrom commented on src/lando/headless_api/tests/test_automation_api.py:
```diff
@@ -0,0 +1,427 @@
+
+@pytest.mark.django_db
+def test_automation_job_add_commit_success(
+    hg_server, hg_clone, hg_automation_worker, repo_mc, monkeypatch
```
Might be worth adding test for git repos upfront, too.

cgsheeh commented on src/lando/headless_api/tests/test_automation_api.py:
```diff
@@ -0,0 +1,427 @@
+from lando.main.scm import SCM_TYPE_HG
+
+
+@pytest.mark.django_db
+def test_auth_missing_user_agent(client, headless_user):
+    user, token = headless_user
+
+    # Create a job and actions
+    job = AutomationJob.objects.create(status=LandingJobStatus.SUBMITTED)
+    AutomationAction.objects.create(
+        job_id=job, action_type="add-commit", data={"content": "test"}, order=0
+    )
+
+    # Fetch job status.
+    response = client.get(
+        f"/api/job/{job.id}",
+        headers={
+            "Authorization": f"Bearer {token}",
+        },
+    )
+
+    assert response.status_code == 401, "Missing `User-Agent` should result in 401."
+    assert response.json() == {"details": "`User-Agent` header is required."}
```
We do want to enforce a `User-Agent` string, but we wanted it in a specific format. I will make that clearer. :)

cgsheeh commented on src/lando/headless_api/api.py:
```diff
@@ -0,0 +1,221 @@
+        return True
+
+
+class MergeOntoAction(Schema):
+    """Merge the current branch into the target commit."""
+
+    action: Literal["merge-onto"]
+    target: str
+    message: str
+
+
+class TagAction(Schema):
+    """Create a new tag with the given name."""
+
+    action: Literal["tag"]
+    name: str
+
+
+class AddBranchAction(Schema):
+    """Create a new branch at the given commit."""
+
+    action: Literal["add-branch"]
+    name: str
+    commit: str
```
Yes, I wanted to get the API up with a single action and we would add further actions in subsequent PRs.

shtrom commented on src/lando/api/legacy/workers/automation_worker.py:
```diff
@@ -0,0 +1,189 @@
+            actions = job.actions.all()
+            for action_row in actions:
+                # Turn the row action into a Pydantic action.
+                action = TypeAdapter(Action).validate_python(action_row.data)
```
I'd suggest hiding this logic within the api logic, and just have something like a `ResolveAction(action_row.data)` here.

zzzeid commented on src/lando/main/models/profile.py:
```diff
@@ -77,6 +77,9 @@ class Meta:
     encrypted_phabricator_api_key = models.BinaryField(default=b"", blank=True)
 
+    # Is the user allowed to make pushes via the automation API.
+    is_automation_user = models.BooleanField(default=False)
```
As mentioned elsewhere, this should be changed to be a permission.

zzzeid reviewed (changes requested):
Did another pass here with a few comments.

I did not do a full pass on `s/lando/headless_api/tests/test_automation_api.py` in this pass, but I will in the next one, given that some of the changes I requested may impact the tests.

cgsheeh commented on src/lando/headless_api/management/commands/start_automation_worker.py:
```diff
@@ -0,0 +1,70 @@
+
+    def handle(self, name: str, **options):
+        """Select a landing worker based on provided argument and start it up."""
+        handlers = {
+            SCM_TYPE_GIT: self.handle_git,
+            SCM_TYPE_HG: self.handle_hg,
+        }
+
+        worker = self.get_worker(name)
+        handlers[worker.scm](worker)
```
We do plan on supporting an hg and git automation worker, so we can move automation to using the new APIs before the official cutover.

cgsheeh commented on src/lando/headless_api/api.py:
```diff
@@ -49,22 +49,22 @@ def authenticate(self, request, token: str) -> str:
 
         try:
-            user = ApiToken.verify_token(token)
+            api_token = ApiToken.verify_token(token)
```
Yes, the `ApiToken` object contains the token prefix and hash, we should consider it sensitive data. I'll rename the variable to `token`. :+1:

cgsheeh replied:
I went with `api_key`.

cgsheeh commented on src/lando/api/legacy/workers/automation_worker.py:
```diff
@@ -0,0 +1,295 @@
+            self.refresh_enabled_repos()
+
+        if self.last_job_finished is False:
+            logger.info("Last job did not complete, sleeping.")
```
This is the same logic used in the landing worker. I'm not sure why we throttle when the previous job wasn't a success, maybe we should revisit that in another bug. :)

cgsheeh commented on src/lando/api/legacy/workers/automation_worker.py:
In the interest of time, I'd like to land this PR in the current state and handle moving the shared code to the `Worker` class as a follow-up. I filed [bug 1958894](https://bugzilla.mozilla.org/show_bug.cgi?id=1958894) for this.

cgsheeh commented on src/lando/headless_api/management/commands/start_automation_worker.py:
I'm going to tackle this at a later time, as [bug 1958894](https://bugzilla.mozilla.org/show_bug.cgi?id=1958894).

cgsheeh commented on src/lando/headless_api/tests/test_automation_api.py:
```diff
@@ -237,8 +237,8 @@ def test_automation_job_create_repo_automation_disabled(
     assert (
-        response.status_code == 403
-    ), "Automation disabled for repo should `403 Forbidden` status."
+        response.status_code == 400
+    ), "Automation disabled for repo should return `400 Bad Request` status."
```
As discussed in our Zoom chat, we will leave this as a 400.

zzzeid replied:
@shtrom technically:
> try again after changing your request

From the API's perspective, this is applicable here, since you can potentially change the repo and resend the request.

shtrom commented on src/lando/headless_api/models/automation_job.py:
```diff
@@ -0,0 +1,149 @@
+    def next_job(cls, repositories: Optional[Iterable[str]] = None) -> Self:
+        """Return the next automation job."""
+        return (
+            cls.objects.filter(status__in=(JobStatus.SUBMITTED, JobStatus.DEFERRED))
```
Question for my understanding: DEFERRED jobs don't look like they are treated any differently from the SUBMITTED job (e.g., no delay). Does this mean we may spin for while picking up and deferring the same job over and over again?

cgsheeh replied:
Yes, but jobs are usually deferred due to tree closures, issues with pushes due to hgmo etc. Deferred usually means "this is safe to retry at some other time".

shtrom commented on src/lando/headless_api/tests/test_automation_api.py:
```diff
@@ -0,0 +1,678 @@
+    assert len(job.landed_commit_id) == 40, "Landed commit ID should be a 40-char SHA."
+
+
+PATCH_GIT_1 = r"""
```
Might be worth moving this to conftest along with the HG patches.

cgsheeh replied:
Will do as a follow-up. :+1:

shtrom reviewed (approved):
LGTM with nits, but I think this can land as is.

shtrom reviewed (changes requested):
Needs pushlog support (sorry)

cgsheeh commented on src/lando/api/legacy/workers/automation_worker.py:
```diff
@@ -0,0 +1,188 @@
+                    JobAction.FAIL,
+                    message=message,
+                )
+                return True  # Do not try again, this is a permanent failure.
```
`run_job` in the `LandingWorker` is still returning a `bool`, so I'd like to keep `run_automation_job` in the `AutomationWorker` also returning a `bool`. `AutomationActionException` is pretty similar to the `PermanentFailureException`/`TemporaryFailureException`, there is room for some clean-up there, but we can do it as a follow-up.

zzzeid commented on src/lando/conftest.py:
```diff
@@ -98,18 +101,14 @@ def _patch(number=0):
+    The diff can  apply on an empty repo to create a known base for application
+    of other patches as part of the tests.
+    """
+    return Path(__file__).parent / "main" / "tests" / "data"
```
This should be using `settings.BASE_DIR`.

cgsheeh replied:
This is pre-existing code that  has been moved to the global conftest. I will update it.

zzzeid commented on src/lando/conftest.py:
```diff
@@ -464,3 +484,15 @@ def create_state_handler(revision, landing_assessment=None):
     return create_state_handler
+
+
+@pytest.fixture
+def treestatus_url():
+    """A string holding the Tree Status base URL."""
+    return "http://treestatus.test"
```
This should be using `settings.TREESTATUS_URL`. It can be overridden in `test_settings` if needed.

cgsheeh replied:
This is pre-existing code that has been moved to the global conftest. I will update it.

zzzeid commented on src/lando/headless_api/models/tokens.py:
```diff
@@ -1,11 +1,14 @@
 
 from lando.main.models.base import BaseModel
 
+API_TOKEN_PREFIX_LENGTH = 8
```
Wondering if this should be set to a smaller length. What advantage do we get of setting it to `8` vs say `2` or `4`? Unlikely to cause a huge performance hit given the limited number of users on the platform.

zzzeid commented on src/lando/headless_api/models/tokens.py:
```diff
@@ -61,6 +64,6 @@ def verify_token(cls, token: str) -> User:
             if not api_token_obj.is_valid:
                 raise ValueError(f"Token {token} has been revoked.")
```
[this](https://github.com/mozilla-conduit/lando/pull/194#discussion_r2007624185) comment was not addressed. It would be better to not differentiate between a non-existing token and a revoked or invalid token. It would be sufficient to say that the token was invalid in both cases.

zzzeid commented on src/lando/api/legacy/workers/automation_worker.py:
```diff
@@ -0,0 +1,188 @@
+                notification parameters.
+        """
+        notify_user_of_landing_failure(
+            job.requester_email, job.landing_job_identifier, job.error, job.id
```
This comment is unaddressed. Also, `AutomationJob` does not have a `landing_job_identifier` attribute, so this will likely fail?

zzzeid reviewed (approved):
lgtm!
